src/cargo/core/compiler/standard_lib.rs

```diff
@@ -38,19 +39,43 @@ pub fn resolve_std<'cfg>(
     crates: &[String],
 ) -> CargoResult<(PackageSet<'cfg>, Resolve, ResolvedFeatures)> {
     let src_path = detect_sysroot_src_path(target_data)?;
-    let to_patch = [
+    let special_std_packages = [
         "rustc-std-workspace-core",
         "rustc-std-workspace-alloc",
         "rustc-std-workspace-std",
     ];
-    let patches = to_patch
+
+    // Specially std packages should be pulled from `library`.
+    let mut patches = special_std_packages
         .iter()
         .map(|&name| {
             let source_path = SourceId::for_path(&src_path.join("library").join(name))?;
             let dep = Dependency::parse_no_deprecated(name, None, source_path)?;
             Ok(dep)
         })
         .collect::<CargoResult<Vec<_>>>()?;
+
+    // If rust-src contains a "vendor" directory, then patch in all the crates it contains.
+    let vendor_path = src_path.join("vendor");
+    if let Ok(vendor_dir) = fs::read_dir(vendor_path) {
+        for entry in vendor_dir {
+            let entry = entry?;
+            let name = entry
+                .file_name()
+                .into_string()
+                .map_err(|_| anyhow::anyhow!("package name wasn't utf8"))?;
+
+            // Ignore the rustc-std-workspace crates.
+            if name.starts_with("rustc-std-workspace-") {
```

Comment:
Could the comment be expanded here about how we don't want to register two patches and the vendors here are for the empty crates.io versions?

Reply:
I'd be curious if we should be doing something nicer when *building* the vendor. Should we delete these? Or make them symlinks? Not sure what's desired for the other use cases like what @matklad was apparently thinking of.

Reply:
For IDEs, it doesn't matter exactly how this all is setup, if we can get some reasonable `cargo metadata` output somehow (medium/long term, this is not an acutely pressing issue).

A naively simplest setup *to consume* would be if all the crates in `rust-src` were hard-vendored. That is, if they only used path dependencies to other crates in `/library` directory.

Reply:
I think ideally we just wouldn't include these in the vendor (maybe manually delete them?), but either way is fine by me.

Reply:
I think the new logic I have that just looks for `rustc-std-workspace-` and remaps them to `library/` is something that IDEs and other tools can easily copy. Although obviously it would be ideal if they didn't have to.